KAFKA-4218: Enable access to key in ValueTransformer and ValueMapper

jeyhunkarimov opened the pull request:
This PR is the partial implementation for KIP-149. As the discussion for this KIP is still ongoing, I made a PR on the "safe" portions of the KIP (so that it can be included in the next release) which are 1) `ValueMapperWithKey`, 2) `ValueTransformerWithKeySupplier`, and 3) `ValueTransformerWithKey`.

jeyhunkarimov committed:
Submit the first version of KIP-149

Fix indent errors

jeyhunkarimov commented:
Retest this please

jeyhunkarimov commented:
ping @guozhangwang @mjsax @dguy @bbejeck

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -176,11 +181,49 @@
+     * The example below counts the number of token of the value string.
+     * <pre>{@code
+     * KStream<String, String> inputStream = builder.stream("topic");
+     * KStream<String, Integer> outputStream = inputStream.mapValues(new ValueMapper<String, Integer> {
```
Example should use `ValueMapperWithKey` and also use the key somehow?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -260,11 +306,55 @@
+     * The example below splits input records {@code <null:String>} containing sentences as values into their words.
+     * <pre>{@code
+     * KStream<byte[], String> inputStream = builder.stream("topic");
+     * KStream<byte[], String> outputStream = inputStream.flatMapValues(new ValueMapper<String, Iterable<String>> {
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1012,11 +1103,84 @@ void to(final String topic,
+     * // register store
+     * builder.addStore(myStore);
+     *
+     * KStream outputStream = inputStream.transformValues(new ValueTransformerSupplier() { ... }, "myValueTransformState");
```
once more

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -377,6 +377,41 @@
 
+    /**
+     * Create a new {@code KTable} by transforming the value of each record in this {@code KTable} into a new value
+     * (with possible new type)in the new {@code KTable}.
```
nit: missing space after `)` (it's a c&p error -- please fix in L348, too)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -377,6 +377,41 @@
+     * The example below counts the number of token of the value string.
+     * <pre>{@code
+     * KTable<String, String> inputTable = builder.table("topic");
+     * KTable<String, Integer> outputTable = inputTable.mapValue(new ValueMapperWithKey<String, String, Integer> {
```
Should we use the key in the code example?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -421,6 +456,51 @@
 
+    /**
+     * Create a new {@code KTable} by transforming the value of each record in this {@code KTable} into a new value
+     * (with possible new type)in the new {@code KTable}.
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -421,6 +456,51 @@
+     * The example below counts the number of token of the value string.
+     * <pre>{@code
+     * KTable<String, String> inputTable = builder.table("topic");
+     * KTable<String, Integer> outputTable = inputTable.mapValue(new ValueMapperWithKey<String, String, Integer> {
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:
```diff
@@ -469,6 +549,55 @@
+     * @deprecated use {@link #mapValues(ValueMapper, Materialized) mapValues(mapper, Materialized.as(queryableStoreName).withValueSerde(valueSerde))}
+     */
+    @Deprecated
+    <VR> KTable<K, VR> mapValues(final ValueMapperWithKey<? super K, ? super V, ? extends VR> mapper, final Serde<VR> valueSerde, final String queryableStoreName);
```
We should not introduce any deprecates methods. The new `ValueMapperWithKey` should be only added to the new API.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerSupplier.java:
```diff
@@ -23,7 +23,9 @@
  * @param <V>  value type
  * @param <VR> transformed value type
  * @see ValueTransformer
+ * @see ValueTransformerWithKey
```
replace with or add `ValueTransformerWithKeySupplier`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKey.java:
```diff
@@ -0,0 +1,100 @@
+ * Note that the key is read-only and should not be modified, as this can lead to corrupt partitioning.
+ * If {@code ValueTransformerWithKey} is applied to a {@link KeyValue} pair record the record's key is preserved.
+ * <p>
+ * Use {@link ValueTransformerSupplier} to provide new instances of {@code ValueTransformer} to Kafka Stream's runtime.
```
update `withKey` twice here

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKey.java:
```diff
@@ -0,0 +1,100 @@
+ * @param <K>  key type
+ * @param <V>  value type
+ * @param <VR> transformed value type
+ * @see ValueTransformerSupplier
```
replace with  `ValueTransformer`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKey.java:
```diff
@@ -0,0 +1,100 @@
+    void init(final ProcessorContext context);
+
+    /**
+     * Transform the given value to a new value.
```
maybe: `Transform the given [key and ]value to a new value.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueMapperWithKey.java:
```diff
@@ -0,0 +1,52 @@
+public interface ValueMapperWithKey<K, V, VR> {
+
+    /**
+     * Map the given value to a new value.
```
maybe: `Map the given [key and ]value to a new value.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKey.java:
```diff
@@ -0,0 +1,100 @@
+     * @param value the value to be transformed
+     * @return the new value
+     */
+    VR transform(final K key, final V value);
```
maybe rename `key` to `readOnlyKey` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueMapperWithKey.java:
```diff
@@ -0,0 +1,52 @@
+     * @param value the value to be mapped
+     * @return the new value
+     */
+    VR apply(final K key, final V value);
```
maybe rename `key` to `readOnlyKey` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformerWithKeySupplier.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * A {@code ValueTransformerWithKeySupplier} interface which can create one or more {@link ValueTransformerWithKey}
+ * instances.
+ * Note that the key is read-only and should not be modified, as this can lead to corrupt partitioning.
```
Do we need this sentence here?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:
```diff
@@ -111,5 +119,44 @@ public R apply(T2 value2, T1 value1) {
+        Objects.requireNonNull(valueMapper, "valueMapper can't be null");
+        return new ValueMapperWithKey<K, V, VR>() {
+            @Override
+            public VR apply(K key, V value) {
```
nit: add `final` to both parameters (same below)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamMapValues.java:
```diff
@@ -37,7 +37,7 @@ public KStreamMapValues(ValueMapper<V, V1> mapper) {
         @Override
         public void process(final K key, final V value) {
-            V1 newValue = mapper.apply(value);
+            V1 newValue = mapper.apply(key, value);
```
nit: add `final` to `V1 newValue`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValues.java:
```diff
@@ -168,16 +168,12 @@ public long timestamp() {
-                throw new StreamsException("ValueTransformer#punctuate must return null.");
-            }
-        }
+        public void punctuate(long timestamp) {}
```
We cannot just remove this... For backward compatibility, we need to be able to trigger `punctuate` for a `ValueTransformer`.

Thus, I would suggest to use a `InternalValueTransformerWithKey extends ValueTransformerWithKey` here, and add a `getValueTransformer` -- if the uses passes a `ValueTransformerWithKeySupplier`, `getValueTransformer` returns `null` -- if the user passes a `ValueTransformerSuppliers`, we wrap it in `withKey` and this wrapper does return the original one `ValueTransformer`.

Thus, here we only all `valueTransformer.punctuate(timestamp)` if we get an `valueTransformer` back.

Maybe some other helpers/wrappers are needed.

Seems we have a test gap here... should have detected this by a failing test. Not sure if it's worth adding one now (add a test for a deprecated API?) \cc @guozhangwang @dguy @bbejeck

dguy replied:
I think it is probably worth adding as even if it is deprecated it is still supported

guozhangwang replied:
+1, agree with @mjsax and @dguy

bbejeck replied:
+1 agree as well

jeyhunkarimov replied:
@mjsax I have some doubts about the solution you proposed. Here, independent of user's using  `ValueTransformerWithKeySupplier` or `ValueTransformerSupplier`, `KStreamTransformValues` class always gets `ValueTransformerWithKeySupplier`. The reason is that we always wrap `ValueTransformerSupplier` into `ValueTransformerWithKeySupplier` and pass and the latter one to `KStreamTransformValues`. Moreover, `ValueTransformerWithKeySupplier` and `ValueTransformerSupplier` are separate interfaces, they don't extend from each other. 
So, I think the most feasible solution for that is to include deprecated method (`punctuate()`) in `ValueTransformerWithKeySupplier`.

mjsax replied:
Because `ValueTransformerWithKeySupplier` is a public interface, we should try to find a solution that does not add a deprecated method to this new interface. If my proposal doesn't work, I am sure there is another solution (using sub-classing etc) to do some internal re-directs to make it work.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -295,6 +307,13 @@ String internalStoreName() {
+    }
+
+    @Override
+    public <VR> KTable<K, VR> mapValues(final ValueMapperWithKey<? super K, ? super V, ? extends VR> mapper,
```
as mentioned above: remove this overlaod

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -269,24 +270,35 @@ String internalStoreName() {
         final KTableProcessorSupplier<K, V, VR> processorSupplier = new KTableMapValues<>(this,
-                                                                                          mapper,
-                                                                                          materializedInternal.storeName());
+                mapper,
```
either move `this` to next line, or fix indention.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -269,24 +270,35 @@ String internalStoreName() {
         builder.internalTopologyBuilder.addStateStore(new KeyValueStoreMaterializer<>(materializedInternal)
-                                                              .materialize(),
-                                                      name);
+                        .materialize(),
```
nit:
```
builder.internalTopologyBuilder.addStateStore(
    new KeyValueStoreMaterializer<>(materializedInternal).materialize(),
    name);
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapValuesTest.java:
```diff
@@ -74,4 +75,43 @@ public void testFlatMapValues() {
+        ValueMapperWithKey<Integer, Number, Iterable<String>> mapper =
+                new ValueMapperWithKey<Integer, Number, Iterable<String>>() {
+            @Override
+            public Iterable<String> apply(Integer key, Number value) {
```
nit: add `final` to parameters

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapValuesTest.java:
```diff
@@ -74,4 +75,43 @@ public void testFlatMapValues() {
+        stream.flatMapValues(mapper).process(processor);
+
+        driver.setUp(builder);
+        for (int expectedKey : expectedKeys) {
```
nit: add `final`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapValuesTest.java:
```diff
@@ -74,4 +75,43 @@ public void testFlatMapValues() {
+
+        final int[] expectedKeys = {0, 1, 2, 3};
+
+        KStream<Integer, Integer> stream;
```
nit: add `final` and put into single liner (same next line)

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapValuesTest.java:
```diff
@@ -74,4 +75,43 @@ public void testFlatMapValues() {
+
+        for (int i = 0; i < expected.length; i++) {
+            assertEquals(expected[i], processor.processed.get(i));
+        }
```
Form L109 on: can be done with a single `assert` that passed in both arrays.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapValuesTest.java:
```diff
@@ -74,4 +75,43 @@ public void testFlatMapValues() {
+
+
+    @Test
+    public void testFlatMapValuesWithKeys() {
```
We should add similar test for mapValuesWithKeys, and ValueTransformerWithKeys, too.

mjsax reviewed (commented):
Thanks for updating the PR. Couple of comments.

A general thought: what should our pattern for using `@see` be? I think, we don't should to cross link too much -- but also not sure what the boarder line / rule should be.

jeyhunkarimov committed:
Add internal value transformer interface

Fix bugs related with ValueTransformer conversion

Fix indentation errors

Add final to apply method parameters

jeyhunkarimov commented:
I added `InternalInternalValueTransformerWithKey` class which can be thought of an IR of `ValueTransformerWithKey` and `ValueTransformer`. That is, both instances are converted to `InternalInternalValueTransformerWithKey` and passed to `KStreamTransformValues`. WDYT?

mjsax commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/wordcount/WordCountDemo.java:
```diff
@@ -62,7 +62,7 @@ public static void main(String[] args) throws Exception {
 
         KStream<String, String> source = builder.stream("streams-plaintext-input");
-
+Arrays.asList(builder);
```
Some left over?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValuesTest.java:
```diff
@@ -86,22 +90,82 @@ public void close() {
+                return new InternalValueTransformerWithKey<Integer, Integer, Integer>() {
+                    @Override
+                    public Integer punctuate(long timestamp) {
+                        return null;
```
This should return `-1` as old `BadValueTransformer`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValuesTest.java:
```diff
@@ -128,16 +192,9 @@ public void shouldNotAllowValueTransformerToCallInternalProcessorContextMethods(
-            fail("should not allow ValueTransformer#puntuate() to return not-null value");
-        } catch (final StreamsException e) {
-            // expected
-        }
```
This should be kept (cf. comment above)

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValuesTest.java:
```diff
@@ -146,7 +203,7 @@ public void init(ProcessorContext context) {
 
         @Override
-        public Integer transform(Integer value) {
+        public Integer transform(Integer key, Integer value) {
```
nit: add `final`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValuesTest.java:
```diff
@@ -86,22 +90,82 @@ public void close() {
+                    }
+
+                    @Override
+                    public Integer transform(Integer readOnlyKey, Integer value) {
```
nit: add `final`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValuesTest.java:
```diff
@@ -86,22 +90,82 @@ public void close() {
+                    }
+
+                    @Override
+                    public void init(ProcessorContext context) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamMapValues.java:
```diff
@@ -16,16 +16,16 @@
+    private final ValueMapperWithKey<K, V, V1> mapper;
 
-    public KStreamMapValues(ValueMapper<V, V1> mapper) {
+    public KStreamMapValues(ValueMapperWithKey<K, V, V1> mapper) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -176,11 +181,49 @@
+     * @see #transformValues(ValueTransformerSupplier, String...)
+     * @see #transformValues(ValueTransformerWithKeySupplier, String...)
+     */
+    <VR> KStream<K, VR> mapValues(ValueMapperWithKey<? super K, ? super V, ? extends VR> mapper);
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:
```diff
@@ -111,5 +120,75 @@ public R apply(T2 value2, T1 value1) {
+                return new InternalValueTransformerWithKey<K, V, VR>() {
+                    @Override
+                    public VR punctuate(final long timestamp) {
+                        return null;
```
I am wondering if we should `throw` here as this should never be called. Or do I miss anything?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapValues.java:
```diff
@@ -16,16 +16,16 @@
+    private final ValueMapperWithKey<? super K, ? super V, ? extends Iterable<? extends V1>> mapper;
 
-    KStreamFlatMapValues(ValueMapper<? super V, ? extends Iterable<? extends V1>> mapper) {
+    KStreamFlatMapValues(ValueMapperWithKey<? super K, ? super V, ? extends Iterable<? extends V1>> mapper) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapValues.java:
```diff
@@ -37,7 +37,7 @@
         @Override
         public void process(K key, V value) {
-            Iterable<? extends V1> newValues = mapper.apply(value);
+            Iterable<? extends V1> newValues = mapper.apply(key, value);
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -514,9 +527,22 @@ private void to(final String topic, final ProducedInternal<K, V> produced) {
+
+    private <VR> KStream<K, VR> transformValues(final InternalValueTransformerWithKeySupplier<? super K, ? super V, ? extends VR> internalValueTransformerWithKeySupplier,
+                                               final String... stateStoreNames) {
+        String name = builder.newProcessorName(TRANSFORMVALUES_NAME);
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -514,9 +527,22 @@ private void to(final String topic, final ProducedInternal<K, V> produced) {
+    }
+
+    private <VR> KStream<K, VR> transformValues(final InternalValueTransformerWithKeySupplier<? super K, ? super V, ? extends VR> internalValueTransformerWithKeySupplier,
+                                               final String... stateStoreNames) {
```
nit: fit indention

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValues.java:
```diff
@@ -36,9 +36,9 @@
+    private final InternalValueTransformerWithKeySupplier<K, V, R> valueTransformerSupplier;
 
-    public KStreamTransformValues(ValueTransformerSupplier<V, R> valueTransformerSupplier) {
+    public KStreamTransformValues(InternalValueTransformerWithKeySupplier<K, V, R> valueTransformerSupplier) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValues.java:
```diff
@@ -49,10 +49,10 @@ public KStreamTransformValues(ValueTransformerSupplier<V, R> valueTransformerSup
         private ProcessorContext context;
 
-        public KStreamTransformValuesProcessor(ValueTransformer<V, R> valueTransformer) {
+        public KStreamTransformValuesProcessor(InternalValueTransformerWithKey<K, V, R> valueTransformer) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -543,9 +553,9 @@ public void to(final Serde<K> keySerde,
+        builder.internalTopologyBuilder.addProcessor(name, new KStreamMapValues<K, Change<V>, V>(new ValueMapperWithKey<K, Change<V>, V>() {
             @Override
-            public V apply(Change<V> change) {
+            public V apply(K key, Change<V> change) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java:
```diff
@@ -65,11 +65,11 @@ public void enableSendingOldValues() {
     }
 
-    private V1 computeValue(V value) {
+    private V1 computeValue(K key, V value) {
```
nit: add `final`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableMapValues.java:
```diff
@@ -91,8 +91,8 @@ public void init(ProcessorContext context) {
-            V1 newValue = computeValue(change.newValue);
-            V1 oldValue = sendOldValues ? computeValue(change.oldValue) : null;
+            V1 newValue = computeValue(key, change.newValue);
+            V1 oldValue = sendOldValues ? computeValue(key, change.oldValue) : null;
```
nit: add `final` (both lines)

mjsax reviewed (commented):
Thanks for the update. Mostly nits.

jeyhunkarimov committed:
Fix nits

jeyhunkarimov commented:
Thanks for the review. I submitted fixed PR.

bbejeck commented:
retest this please

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1012,11 +1108,84 @@ void to(final String topic,
+     * In order to assign a state, the state must be created and registered beforehand:
+     * <pre>{@code
+     * // create store
+     * StateStoreSupplier myStore = Stores.create("myValueTransformState")
```
nit: the `StateStoreSupplier` example is deprecated API maybe change to something like
```
StoreBuilder<KeyValueStore<String,String>> keyValueStoreBuilder = 
Stores.keyValueStoreBuilder(Stores.persistentKeyValueStore("myValueTransformState"),
                             Serdes.String(),
                             Serdes.String());

builder.addStateStore(keyValueStore);
```
EDIT: I understand this is how other examples are presented in javadoc, probably worth JIRA (KAFKA-6411)  to update to use new API in examples.

jeyhunkarimov replied:
Thanks a lot for the feedback. Fixed.

jeyhunkarimov committed:
Fix outdated javadocs

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/InternalValueTransformerWithKey.java:
```diff
@@ -0,0 +1,22 @@
+/*
```
This class and the next one should be in `.../kstream/internals` right?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -295,11 +309,7 @@ String internalStoreName() {
-            storeSupplier = keyValueStore(this.keySerde, valueSerde, queryableStoreName);
-        }
-        return doMapValues(mapper, valueSerde, storeSupplier);
+        return mapValues(withKey(mapper), Materialized.<K, V1, KeyValueStore<Bytes, byte[]>>as(queryableStoreName).withValueSerde(valueSerde));
```
Seems we did not pass in the `this.keySerde` here any more?

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValuesTest.java:
```diff
@@ -86,22 +90,82 @@ public void close() {
+                return new InternalValueTransformerWithKey<Integer, Integer, Integer>() {
+                    @Override
+                    public Integer punctuate(long timestamp) {
+                        return -1;
```
We should also be able to just throw here since it is never expected, right?

guozhangwang reviewed (commented):
@jeyhunkarimov Thanks for the updated PR! I made a pass over it and only have two minor comments, otherwise LGTM!

jeyhunkarimov committed:
Fix minor issues

jeyhunkarimov commented:
Thanks, @guozhangwang for your review. I submitted a fix.

guozhangwang commented:
LGTM!

guozhangwang commented:
Merged to trunk. @jeyhunkarimov Thank you for your patience and the great effort!
